#!/usr/bin/perl

##### perl all_new_transcript.pl refernce.gtf cuffcompare.tracking cuffcompare.combined.gtf

@ARGV = qw { Homo_sapiens.GRCh38.79.gtf Lnchsvnovel.tracking Lnchsvnovel.combined.gtf };
open ref_gtf, "<$ARGV[0]";  #### reference gtf file;
while (<ref_gtf>) {
		@each_feature=split (/\t/, $_);
		$each_feature[8]=~ /gene_id \"(.*?)\"\; transcript_id \"(.*?)\";/;
		$ref_trancript_gene {$2}= $1;
}


open tracking, "<$ARGV[1]";  ####cuffcompare tracking file
while (<tracking>) {
	@eachline =split (/\t/, $_);
	push @{$loc_class_code{$eachline[1]}}, $eachline[3];
	push @{$gene_transcript_tracking{$eachline[1]}}, $eachline[0], 
	
}
close tracking;



open  combined_gtf, "<$ARGV[2]";  ###cuffcompare combined gtf file
while (<combined_gtf>) {
		@each_feature=split (/\t/, $_);
		$each_feature[8]=~ /gene_id \"(.*?)\"\; transcript_id \"(.*?)\";/;
		push @{$combined_gtf_transcript_exon_information {$2}}, $_;
		push @{$gene_transcript{$1}}, $2;
		$transcript_exon_number{$2}++;
		$transcript_gene{$2} = $1;
		$transcript_length{$2} = $transcript_length{$2}+$each_feature[4]- $each_feature[3]+1;

}
close combined_gtf;





open tracking, "<$ARGV[1]";   ##tracking generated by cuffcompare
while (<tracking>) {
	chomp $_;
	@each_transcript_feature =split (/\t/, $_);
	$TCONS= shift @each_transcript_feature;
	$XLOC = shift @each_transcript_feature;
	$name_gene_transcript = shift @each_transcript_feature;
	($ref_name_gene, $ref_transcript_id) =split (/\|/, $name_gene_transcript);
	$ref_gene_id = $ref_trancript_gene{$ref_transcript_id};
	$class_code = shift @each_transcript_feature;
	$x_length_transcript = $transcript_length{$TCONS};




if ($x_length_transcript >199) {

	if (($class_code eq "c") or ($class_code eq "j") or ($class_code eq "o") or ($class_code eq "e")) {
		$exon_number_this_transcript= $transcript_exon_number{$TCONS};
		if ($exon_number_this_transcript>1) {### if exon number of this transcript >1 
		
				for ($i=0;$i<@each_transcript_feature;$i++) {
					$cufflinks_transcript_FPKM[$i]=0;

					#q1:ENSMUSG00000090025|ENSMUST00000160944|0|0.000000|0.000000|0.000000|0.000000|501
					if ($each_transcript_feature[$i] ne "-") {
						if ($each_transcript_feature[$i] =~ /\,/) {
							@each_x= split (/\,/, $each_transcript_feature[$i]);
							foreach $mmm (@each_x) {
								($xxx, $yyy)= split (/\:/, $mmm);
								@each_information =split (/\|/, $yyy);
								$cufflinks_transcript_FPKM[$i] = $cufflinks_transcript_FPKM[$i]+ $each_information[3];
							}

						}
						else {
							($xxx, $yyy)= split (/\:/, $each_transcript_feature[$i]);
							@each_information =split (/\|/, $yyy);
							$cufflinks_transcript_FPKM[$i]= $each_information[3];
						}

					}
					else {
						$cufflinks_transcript_FPKM[$i]=0;
					}

				}
				
				$if_exist_fpkm1= if_large_1(@cufflinks_transcript_FPKM);   ### at least one tissue FPKM>1
		
				if ($if_exist_fpkm1 == 1) {

						for ($i=0;$i<@ {$combined_gtf_transcript_exon_information{$TCONS}};$i++) {
							$each_combined_gtf_exon_information = $combined_gtf_transcript_exon_information{$TCONS} -> [$i];
							$each_combined_gtf_exon_information =~ s/$XLOC/$ref_gene_id/g;
							push @all_new_transcript, $each_combined_gtf_exon_information;
						}
				}
		}
	}

}

}










%seen=();
my @unique3 = grep { ! $seen{$_} ++ } @all_new_transcript; 


open result_all_new_transcript, ">all_new_transcript.gtf";
foreach $i (@unique3) {
	print result_all_new_transcript "$i";
}



sub if_large_1 {
	my @input= @_;
	$numx=0;
	foreach $value (@input) {
		if ($value >1) {
			$numx++;
			last;
		}
	}
	if ($numx == 0) {return 0;}
	else {return 1;}
}



